File: packages/runtime-ts/src/services/telemetry.ts
/**
 * OpenTelemetry integration for GlassOps Runtime.
 * Provides tracing and metrics for governance operations.
 */

import { trace, SpanStatusCode, Span, context } from "@opentelemetry/api";
import { NodeSDK } from "@opentelemetry/sdk-node";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-http";
import {
  ATTR_SERVICE_NAME,
  ATTR_SERVICE_VERSION,
} from "@opentelemetry/semantic-conventions";
import { Resource } from "@opentelemetry/resources";

let sdk: NodeSDK | null = null;
const TRACER_NAME = "glassops-runtime";

/**
 * Initialize OpenTelemetry SDK.
 * Only initializes if OTEL_EXPORTER_OTLP_ENDPOINT is set.
 */
export async function initTelemetry(
  serviceName: string = "glassops-runtime",
  serviceVersion: string = "1.0.0",
): Promise<void> {
  const endpoint = process.env.OTEL_EXPORTER_OTLP_ENDPOINT;

  if (!endpoint) {
    // No endpoint configured, telemetry disabled
    return;
  }

  try {
    const resource = new Resource({
      [ATTR_SERVICE_NAME]: serviceName,
      [ATTR_SERVICE_VERSION]: serviceVersion,
    });

    const headersRaw = process.env.OTEL_EXPORTER_OTLP_HEADERS;
    const headers: Record<string, string> = {};

    if (headersRaw) {
      // Parse "Header=Value,Header2=Value2" format
      headersRaw.split(",").forEach((header) => {
        const [key, value] = header.split("=");
        if (key && value) {
          headers[key.trim()] = value.trim();
        }
      });
    }

    const traceExporter = new OTLPTraceExporter({
      url: `${endpoint}/v1/traces`,
      headers,
    });

    sdk = new NodeSDK({
      resource,
      traceExporter,
    });

    await sdk.start();
  } catch (error) {
    // Telemetry initialization failure should not block execution
    console.warn(
      `[Telemetry] Failed to initialize: ${error instanceof Error ? error.message : String(error)}`,
    );
  }
}

/**
 * Shutdown telemetry gracefully.
 */
export async function shutdown(): Promise<void> {
  if (sdk) {
    try {
      await sdk.shutdown();
    } catch {
      // Ignore shutdown errors
    }
    sdk = null;
  }
}

/**
 * Execute an async function within a traced span.
 */
export async function withSpan<T>(
  name: string,
  fn: (span: Span) => Promise<T>,
  attributes?: Record<string, string | number | boolean>,
): Promise<T> {
  const tracer = trace.getTracer(TRACER_NAME);

  return tracer.startActiveSpan(name, async (span) => {
    if (attributes) {
      span.setAttributes(attributes);
    }

    try {
      const result = await fn(span);
      span.setStatus({ code: SpanStatusCode.OK });
      return result;
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      span.setStatus({ code: SpanStatusCode.ERROR, message });
      span.recordException(error instanceof Error ? error : new Error(message));
      throw error;
    } finally {
      span.end();
    }
  });
}

/**
 * Get the current active span (if any).
 */
export function getCurrentSpan(): Span | undefined {
  return trace.getSpan(context.active());
}

/**
 * Add an event to the current span.
 */
export function addSpanEvent(
  name: string,
  attributes?: Record<string, string | number | boolean>,
): void {
  const span = getCurrentSpan();
  if (span) {
    span.addEvent(name, attributes);
  }
}

File: packages/runtime/internal/telemetry/telemetry.go
// Package telemetry provides OpenTelemetry integration for tracing.
package telemetry

import (
	"context"
	"os"
	"strings"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.24.0"
	"go.opentelemetry.io/otel/trace"
)

const tracerName = "glassops-runtime"

var (
	tracerProvider *sdktrace.TracerProvider
	tracer         trace.Tracer
)

// Init initializes OpenTelemetry SDK.
// Only initializes if OTEL_EXPORTER_OTLP_ENDPOINT is set.
func Init(ctx context.Context, serviceName, serviceVersion string) error {
	endpoint := os.Getenv("OTEL_EXPORTER_OTLP_ENDPOINT")
	if endpoint == "" {
		return nil // Telemetry disabled
	}

	// Parse headers from environment
	headersRaw := os.Getenv("OTEL_EXPORTER_OTLP_HEADERS")
	headers := make(map[string]string)
	if headersRaw != "" {
		for _, header := range strings.Split(headersRaw, ",") {
			parts := strings.SplitN(header, "=", 2)
			if len(parts) == 2 {
				headers[strings.TrimSpace(parts[0])] = strings.TrimSpace(parts[1])
			}
		}
	}

	exporter, err := otlptracehttp.New(ctx,
		otlptracehttp.WithEndpoint(strings.TrimPrefix(endpoint, "https://")),
		otlptracehttp.WithHeaders(headers),
	)
	if err != nil {
		return err
	}

	res, err := resource.New(ctx,
		resource.WithAttributes(
			semconv.ServiceNameKey.String(serviceName),
			semconv.ServiceVersionKey.String(serviceVersion),
		),
	)
	if err != nil {
		return err
	}

	tracerProvider = sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter),
		sdktrace.WithResource(res),
	)

	otel.SetTracerProvider(tracerProvider)
	tracer = otel.Tracer(tracerName)

	return nil
}

// Shutdown gracefully shuts down the telemetry provider.
func Shutdown(ctx context.Context) error {
	if tracerProvider != nil {
		return tracerProvider.Shutdown(ctx)
	}
	return nil
}

// WithSpan executes a function within a traced span.
func WithSpan[T any](ctx context.Context, name string, fn func(context.Context, trace.Span) (T, error), attrs ...attribute.KeyValue) (T, error) {
	if tracer == nil {
		tracer = otel.Tracer(tracerName)
	}

	ctx, span := tracer.Start(ctx, name)
	defer span.End()

	if len(attrs) > 0 {
		span.SetAttributes(attrs...)
	}

	result, err := fn(ctx, span)
	if err != nil {
		span.SetStatus(codes.Error, err.Error())
		span.RecordError(err)
	} else {
		span.SetStatus(codes.Ok, "")
	}

	return result, err
}

// GetCurrentSpan returns the current active span from context.
func GetCurrentSpan(ctx context.Context) trace.Span {
	return trace.SpanFromContext(ctx)
}

// AddSpanEvent adds an event to the current span.
func AddSpanEvent(ctx context.Context, name string, attrs ...attribute.KeyValue) {
	span := trace.SpanFromContext(ctx)
	if span != nil {
		span.AddEvent(name, trace.WithAttributes(attrs...))
	}
}
